File: frontend/src/lib/tiers.ts
export type Tier = "free" | "pro" | "elite";

export const TIER_FEATURES: Record<Tier, Record<string, boolean | number>> = {
  free: {
    web_search: false,
    adaptation: false,
    collective: false,
    coach_chat: false,
    sport_specific: false,
    max_plans_per_month: 1,
    max_mesocycle_weeks: 4,
  },
  pro: {
    web_search: true,
    adaptation: true,
    collective: true,
    coach_chat: false,
    sport_specific: false,
    max_plans_per_month: -1,
    max_mesocycle_weeks: 12,
  },
  elite: {
    web_search: true,
    adaptation: true,
    collective: true,
    coach_chat: true,
    sport_specific: true,
    max_plans_per_month: -1,
    max_mesocycle_weeks: 16,
  },
};

export function canUse(tier: Tier, feature: string): boolean {
  return !!TIER_FEATURES[tier]?.[feature];
}

export interface TierDisplay {
  name: string;
  price: string;
  features: string[];
  highlight?: boolean;
}

export const TIER_DISPLAY: Record<Tier, TierDisplay> = {
  free: {
    name: "Free",
    price: "$0",
    features: [
      "1 plan per month",
      "4-week training blocks",
      "Certified trainer persona",
      "Based on training principles",
    ],
  },
  pro: {
    name: "Pro",
    price: "$19/mo",
    highlight: true,
    features: [
      "Unlimited plans",
      "8-12 week periodized mesocycles",
      "World-class S&C coach persona",
      "Evidence-based research (PubMed)",
      "Weekly plan adaptation",
      "Collective learning insights",
    ],
  },
  elite: {
    name: "Elite",
    price: "$49/mo",
    features: [
      "Everything in Pro",
      "Up to 16-week mesocycles",
      "Sport-specific Olympic-level coach",
      "Deep sport research",
      "Competition peaking model",
      "On-demand adaptation",
      "Coach chat",
    ],
  },
};

File: backend/tiers.py
from datetime import datetime, timedelta, timezone

PERSONAS = {
    "free": (
        "You are a certified personal trainer with solid foundational knowledge. "
        "You give safe, effective programming based on established training principles. "
        "Keep recommendations straightforward and proven."
    ),
    "pro": (
        "You are a world-class strength and conditioning coach with 20+ years of "
        "experience training athletes from recreational to national level. You base "
        "every decision on current sports science literature. Your programming is "
        "precise — specific loads, specific RPE targets, specific progression schemes. "
        "You never say 'moderate weight' or 'appropriate intensity.'"
    ),
    "elite": None,  # dynamically generated via build_elite_persona()
}

SPORT_DEMANDS = {
    "swimming": (
        "shoulder stability and mobility, rotational core power, lat and posterior "
        "chain strength, kick power through hip flexor and quad development, and the "
        "critical balance between dryland training and pool volume. You understand "
        "taper protocols for competition peaks and how to periodize strength work "
        "around high-yardage training blocks."
    ),
    "running": (
        "lower limb strength-endurance, injury-resilient posterior chain, plyometric "
        "capacity, running economy through force production, and the balance between "
        "strength training and running volume. You understand periodization around "
        "race calendars and the taper-to-peak cycle."
    ),
    "powerlifting": (
        "competition squat/bench/deadlift specificity, peaking and attempt selection, "
        "weak point identification, accessory programming for structural balance, and "
        "meet-day preparation. You understand Sheiko, Westside, conjugate, and block "
        "periodization models and when each applies."
    ),
    "crossfit": (
        "concurrent development of strength, gymnastics capacity, and metabolic "
        "conditioning. You understand competitive CrossFit programming, engine "
        "building, skill acquisition prioritization, and how to peak for the Open "
        "or Quarterfinals."
    ),
    "basketball": (
        "explosive power, lateral agility, vertical leap development, injury "
        "prevention for ankles and knees, and in-season load management. You "
        "understand how to maintain strength during a competitive season without "
        "compromising game performance."
    ),
    "soccer": (
        "repeated sprint ability, lower body power, hamstring injury prevention, "
        "aerobic and anaerobic capacity, and in-season load management. You understand "
        "GPS data-informed training and how to balance gym work with pitch sessions."
    ),
    "tennis": (
        "rotational power, shoulder durability (including dominant-side glenohumeral "
        "internal rotation deficit management, GIRD), lateral movement capacity, "
        "anti-rotation core strength, and tournament-schedule periodization. You "
        "understand the unique demands of a sport with no off-season and frequent travel."
    ),
    "mma": (
        "strength-to-weight optimization, grip and neck strength, power endurance, "
        "fight-camp periodization, and evidence-based weight management (prioritizing "
        "gradual longitudinal weight descent over acute dehydration-based cuts to "
        "preserve strength and power into competition). You understand how to "
        "peak strength and conditioning for a specific fight date while managing "
        "sparring load."
    ),
    "cycling": (
        "leg strength-endurance, power-to-weight ratio, core stability for aero "
        "position, and the balance between gym work and on-bike volume. You understand "
        "FTP-based training zones and how strength training transfers to cycling power."
    ),
    "general": (
        "broad physical preparedness, injury resilience, and progressive overload for "
        "lifelong health. You emphasize movement quality and sustainable programming."
    ),
}


def build_elite_persona(sport: str) -> str:
    # Validate sport against known keys; for freetext "Other", sanitize
    import re as _re
    if sport not in SPORT_DEMANDS:
        sport = _re.sub(r"[^a-zA-Z0-9 \-]", "", sport)[:30] or "general"
    demands = SPORT_DEMANDS.get(sport, SPORT_DEMANDS["general"])
    return (
        f"You are an elite {sport} strength and conditioning coach who has "
        f"trained multiple athletes at the Olympic or professional level. "
        f"You understand the specific physical demands of {sport}: {demands} "
        f"Your programming reflects decades of experience at the highest level. "
        f"Every exercise choice, volume prescription, and intensity target is "
        f"justified by how it transfers to {sport} performance. You never "
        f"program generic 'gym bro' training — everything serves the sport."
    )


TIER_FEATURES = {
    "free": {
        "web_search": False,
        "adaptation": False,
        "collective": False,
        "coach_chat": False,
        "sport_specific": False,
        "max_plans_per_month": 1,
        "max_mesocycle_weeks": 4,
    },
    "pro": {
        "web_search": True,
        "adaptation": True,
        "collective": True,
        "coach_chat": False,
        "sport_specific": False,
        "max_plans_per_month": -1,
        "max_mesocycle_weeks": 12,
    },
    "elite": {
        "web_search": True,
        "adaptation": True,
        "collective": True,
        "coach_chat": True,
        "sport_specific": True,
        "max_plans_per_month": -1,
        "max_mesocycle_weeks": 16,
    },
}


def check_feature(user, feature: str) -> bool:
    return TIER_FEATURES[user.tier].get(feature, False)


def check_plan_limit(user, db) -> bool:
    from models.plan import Plan

    limit = TIER_FEATURES[user.tier]["max_plans_per_month"]
    if limit == -1:
        return True
    count = (
        db.query(Plan)
        .filter(
            Plan.user_id == user.id,
            Plan.created_at >= datetime.now(timezone.utc) - timedelta(days=30),
        )
        .count()
    )
    return count < limit
